File: Abstract Games/src/main/kotlin/com/castlefrog/games/asg/HexView.java
package com.castlefrog.games.asg;

import java.util.List;

import android.app.Activity;
import android.graphics.Path;
import android.view.View;
import android.view.MotionEvent;
import android.content.Context;
import android.graphics.drawable.ShapeDrawable;
import android.graphics.drawable.shapes.PathShape;
import android.graphics.Canvas;
import android.graphics.Paint;
import android.graphics.Color;
import android.graphics.PointF;

import com.castlefrog.agl.Arbiter;
import com.castlefrog.agl.Agent;
import com.castlefrog.agl.agents.ExternalAgent;
import com.castlefrog.agl.domains.hex.HexAction;
import com.castlefrog.agl.domains.hex.HexState;
import com.castlefrog.agl.domains.hex.HexSimulator;

public final class HexView extends View {
    private static final int WIN_COLOR = 0xffffffff;
    private static final int PADDING = 10;
    private static final int SCALE = 100;

    private PointF[][] locations_;
    private ShapeDrawable hexagon_;
    /** Distance from center to corner */
    private float hexagonRadius_;
    private float hexagonCRadius_;
    private HexAction selected_;
    private int padding_;

    private List<GameActivity.AgentColor> agentColors_;
    private int[] boardColors_;

    private Paint paint_ = new Paint();
    private Path path_ = new Path();

    private Arbiter<?, ?> arbiter_;

    public HexView(Context context) {
        super(context);

        final float scale = getResources().getDisplayMetrics().density;
        padding_ = (int) (PADDING * scale);
        arbiter_ = ((GameActivity) context).getArbiter();

        //Use an xml resource to recall selected board style and board color
        //SharedPreferences settings = context.getSharedPreferences("hex", Context.MODE_PRIVATE);
        boardColors_ = new int[] {0xffffffff, 0xffaaaaaa};
        agentColors_ = ((GameActivity) context).getAgentColors();

        hexagon_ = new ShapeDrawable(new PathShape(Utils.getHexagon(SCALE), 2 * SCALE, 2 * SCALE * Utils.HEXAGON_SHORT_RADIUS));
        hexagon_.getPaint().setAntiAlias(true);
        hexagon_.getPaint().setStrokeWidth(0.1f * SCALE);

        int size = ((HexSimulator) arbiter_.getWorld()).getState().getBoardSize();
        locations_ = new PointF[size][size];
    }

    @Override
    public boolean onTouchEvent(MotionEvent ev) {
        float x = ev.getX();
        float y = ev.getY();

        View menuWindow = ((Activity) getContext()).findViewById(R.id.menuWindow);

        if (menuWindow.getVisibility() != View.VISIBLE) {
            switch (ev.getAction()) {
            case MotionEvent.ACTION_DOWN:
            case MotionEvent.ACTION_MOVE:
                for (int i = 0; i < locations_.length; i += 1) {
                    for (int j = 0; j < locations_.length; j += 1) {
                        PointF point = locations_[i][j];
                        if (Math.hypot(point.x - x, point.y - y) < hexagonCRadius_) {
                            HexSimulator simulator = (HexSimulator) arbiter_.getWorld();
                            HexState state = simulator.getState();
                            HexAction action = HexAction.valueOf(i , j);
                            int agentTurn = state.getAgentTurn();
                            List<HexAction> legalActions = simulator.getLegalActions(agentTurn);
                            List<Agent> agents = arbiter_.getAgents();
                            if (agents.get(agentTurn) instanceof ExternalAgent && legalActions.contains(action)) {
                                if (selected_ == null || !selected_.equals(action)) {
                                    selected_ = action;
                                    invalidate();
                                }
                                return true;
                            }
                        }
                    }
                }
                if (selected_ != null) {
                    selected_ = null;
                    invalidate();
                }
                break;
            case MotionEvent.ACTION_UP:
                if (selected_ != null) {
                    ((GameActivity) getContext()).inputAction(selected_);
                    selected_ = null;
                    invalidate();
                }
                break;
            default:
                break;
            }
        }
        return true;
    }

    @Override
    public void onSizeChanged(int width, int height, int oldWidth, int oldHeight) {
        super.onSizeChanged(width, height, oldWidth, oldHeight);
        int paddedWidth = width - padding_;
        int paddedHeight = height - padding_;

        HexSimulator simulator = (HexSimulator) arbiter_.getWorld();
        int boardSize = simulator.getState().getBoardSize();

        float maxHexWidth = paddedWidth / (0.75f * boardSize + 0.25f);
        float maxHexHeight = (2 * paddedHeight) / (3 * boardSize - 1);
        hexagonRadius_ = Math.min(maxHexWidth, maxHexHeight / Utils.HEXAGON_SHORT_RADIUS) / 2;
        hexagonCRadius_ = Utils.HEXAGON_SHORT_RADIUS * hexagonRadius_;
        float boardWidth = hexagonRadius_ * (1.5f * boardSize + 0.5f);
        float boardHeight = hexagonCRadius_ * (3 * boardSize - 1);
        float xPadding = (width - boardWidth) / 2;
        float yPadding = (height - boardHeight) / 2;
        float yAdjust = hexagonCRadius_ * (boardSize - 1);

        float x0 = xPadding + hexagonRadius_;
        float y0 = yAdjust + yPadding + (Utils.HEXAGON_SHORT_RADIUS * hexagonRadius_);
        for (int i = 0; i < boardSize; i += 1) {
            for (int j = 0; j < boardSize; j += 1) {
                locations_[i][j] = new PointF(x0 + i * hexagonRadius_ * 1.5f,
                                         y0 - i * hexagonCRadius_ + j * 2 * hexagonCRadius_);
            }
        }
    }

    private PointF getCorner(PointF location,
                             int xChange,
                             int yChange) {
        return new PointF((location.x + xChange * Utils.HEXAGON_HALF_RADIUS * hexagonRadius_),
                          (location.y + yChange * Utils.HEXAGON_SHORT_RADIUS * hexagonRadius_));
    }

    @Override
    public void onDraw(Canvas canvas) {
        super.onDraw(canvas);
        HexSimulator simulator = (HexSimulator) arbiter_.getWorld();
        HexState state = simulator.getState();
        int boardSize = state.getBoardSize();
        byte[][] locations = state.getLocations();

        //draw board outer edges
        paint_.setStyle(Paint.Style.FILL);
        paint_.setColor(agentColors_.get(0).intValue());
        PointF p1 = getCorner(locations_[0][0], -2, 0);
        PointF p2 = locations_[0][boardSize - 1];
        canvas.drawRect(p1.x, p1.y, p2.x, p2.y, paint_);
        p1 = getCorner(locations_[boardSize - 1][0], 2, 0);
        p2 = locations_[boardSize - 1][boardSize - 1];
        canvas.drawRect(p1.x, p1.y, p2.x, p2.y, paint_);

        paint_.setColor(agentColors_.get(1).intValue());
        p1 = getCorner(locations_[0][0], -1, -1);
        p2 = getCorner(locations_[boardSize - 1][0], -1, -1);
        PointF p3 = locations_[boardSize - 1][0];
        PointF p4 = locations_[0][0];
        path_.reset();
        path_.moveTo(p1.x, p1.y);
        path_.lineTo(p2.x, p2.y);
        path_.lineTo(p3.x, p3.y);
        path_.lineTo(p4.x, p4.y);
        path_.close();
        canvas.drawPath(path_, paint_);
        p1 = getCorner(locations_[0][boardSize - 1], 1, 1);
        p2 = getCorner(locations_[boardSize - 1][boardSize - 1], 1, 1);
        p3 = locations_[boardSize - 1][boardSize - 1];
        p4 = locations_[0][boardSize - 1];
        path_.reset();
        path_.moveTo(p1.x, p1.y);
        path_.lineTo(p2.x, p2.y);
        path_.lineTo(p3.x, p3.y);
        path_.lineTo(p4.x, p4.y);
        path_.close();
        canvas.drawPath(path_, paint_);

        //draw board background
        for (int i = 0; i < locations.length; i += 1) {
            for (int j = 0; j < locations[i].length; j += 1) {
                hexagon_.getPaint().setStyle(Paint.Style.FILL);
                PointF point = locations_[i][j];
                hexagon_.setBounds((int) (point.x - hexagonRadius_),
                                   (int) (point.y - hexagonCRadius_),
                                   (int) (point.x + hexagonRadius_),
                                   (int) (point.y + hexagonCRadius_));
                if (locations[i][j] != 0) {
                    hexagon_.getPaint().setColor(agentColors_.get(locations[i][j] - 1).intValue());
                } else {
                    hexagon_.getPaint().setColor(boardColors_[1]);
                }
                hexagon_.draw(canvas);
            }
        }
        //draw winning connection
        List<HexAction> connection = simulator.getWinningConnection();
        for (HexAction location : connection) {
            hexagon_.getPaint().setStyle(Paint.Style.FILL);
            hexagon_.getPaint().setColor(WIN_COLOR);
            PointF point = locations_[location.getX()][location.getY()];
            hexagon_.setBounds((int) (point.x - hexagonRadius_),
                    (int) (point.y - hexagonCRadius_),
                    (int) (point.x + hexagonRadius_),
                    (int) (point.y + hexagonCRadius_));
            hexagon_.draw(canvas);
        }
        //Draw board inner edges
        for (PointF[] row: locations_) {
            for (PointF point: row) {
                hexagon_.getPaint().setStyle(Paint.Style.STROKE);
                hexagon_.getPaint().setColor(boardColors_[0]);
                hexagon_.setBounds((int) (point.x - hexagonRadius_),
                                   (int) (point.y - hexagonCRadius_),
                                   (int) (point.x + hexagonRadius_),
                                   (int) (point.y + hexagonCRadius_));
                hexagon_.draw(canvas);
            }
        }
        //draw selected lines
        if (selected_ != null) {
            int y = selected_.getY();
            int x = selected_.getX();
            paint_.setStyle(Paint.Style.STROKE);
            paint_.setColor(Color.BLACK);
            paint_.setStrokeWidth(4f);
            PointF p1i = locations_[x][0];
            PointF p1f = locations_[x][locations_.length - 1];
            PointF p2i = locations_[0][y];
            PointF p2f = locations_[locations_.length - 1][y];
            canvas.drawLine(p1i.x, p1i.y, p1f.x, p1f.y, paint_);
            canvas.drawLine(p2i.x, p2i.y, p2f.x, p2f.y, paint_);
        }
    }
}

File: Abstract Games/src/main/kotlin/com/castlefrog/games/asg/HexView.kt
package com.castlefrog.games.asg

import android.content.Context
import android.graphics.*
import android.util.AttributeSet
import android.view.MotionEvent
import android.view.View
import com.castlefrog.agl.domains.hex.HexAction
import java.util.ArrayList

public class HexView : View {

    companion object {
        private val DEFAULT_BOARD_SIZE = 5
        private val LINE_WIDTH_RATIO = 0.1f

        public val BACKGROUND_COLOR_VALUE: Int = 0
        public val AGENT1_COLOR_VALUE: Int = 1
        public val AGENT2_COLOR_VALUE: Int = 2
        public val CONNECTION_COLOR_VALUE: Int = 3
    }

    private var locations: MutableList<MutableList<PointF>> = ArrayList()
    private var hexagon: Path? = null
    /** Distance from center to corner  */
    private var hexagonRadius: Float = 0f
    private var hexagonCRadius: Float = 0f
    private var lineWidth: Float = 0f

    private var locationColors: Array<ByteArray> = Array(0, { ByteArray(0) })
    private val paletteColors = ArrayList<Int>()
    private var boardOutlineColor: Int = Color.WHITE
    private var lineColor = Color.BLACK

    private var contentWidth: Int = 0
    private var contentHeight: Int = 0
    private val paint = Paint()
    private val path = Path()

    public var boardSize: Int = 0
        set(value) {
            $boardSize = value
            invalidate()
        }
    public var boardBackgroundColor: Int
        get() = paletteColors.get(BACKGROUND_COLOR_VALUE)
        set(value) {
            paletteColors.set(BACKGROUND_COLOR_VALUE, value)
            invalidate()
        }
    public var agent1Color: Int
        get() = paletteColors.get(AGENT1_COLOR_VALUE)
        set(value) {
            paletteColors.set(AGENT1_COLOR_VALUE, value)
            invalidate()
        }
    public var agent2Color: Int
        get() = paletteColors.get(AGENT2_COLOR_VALUE)
        set(value) {
            paletteColors.set(AGENT2_COLOR_VALUE, value)
            invalidate()
        }
    public var connectionColor: Int
        get() = paletteColors.get(CONNECTION_COLOR_VALUE)
        set(value) {
            paletteColors.set(CONNECTION_COLOR_VALUE, connectionColor)
            invalidate()
        }

    private var selectedHex: Point? = null
    var selectActionListener: SelectActionListener<HexAction> = DummySelectActionListener()

    public constructor(context: Context) : super(context) {
        init(null, 0)
    }

    public constructor(context: Context, attrs: AttributeSet) : super(context, attrs) {
        init(attrs, 0)
    }

    public constructor(context: Context, attrs: AttributeSet, defStyle: Int) : super(context, attrs, defStyle) {
        init(attrs, defStyle)
    }

    private fun init(attrs: AttributeSet?, defStyle: Int) {
        // Load attributes
        val a = getContext().obtainStyledAttributes(attrs, R.styleable.HexView, defStyle, 0)

        boardSize = a.getInt(R.styleable.HexView_boardSize, DEFAULT_BOARD_SIZE)
        val boardBackgroundColor = a.getColor(R.styleable.HexView_boardBackgroundColor, Color.GRAY)
        val agent1Color = a.getColor(R.styleable.HexView_agent1Color, Color.RED)
        val agent2Color = a.getColor(R.styleable.HexView_agent2Color, Color.BLUE)
        val connectionColor = a.getColor(R.styleable.HexView_connectionColor, Color.WHITE)
        boardOutlineColor = a.getColor(R.styleable.HexView_boardOutlineColor, boardOutlineColor)
        lineColor = a.getColor(R.styleable.HexView_lineColor, lineColor)

        a.recycle()

        paletteColors.add(boardBackgroundColor)
        paletteColors.add(agent1Color)
        paletteColors.add(agent2Color)
        paletteColors.add(connectionColor)

        for (i in 1..boardSize) {
            val temp = ArrayList<PointF>()
            for (j in 1..boardSize) {
                temp.add(null)
            }
            locations.add(temp)
        }
        locationColors = Array(boardSize, { ByteArray(boardSize) })
    }

    override fun onTouchEvent(ev: MotionEvent): Boolean {
        if (isEnabled()) {
            val x = ev.getX()
            val y = ev.getY()
            when (ev.getAction()) {
                MotionEvent.ACTION_DOWN, MotionEvent.ACTION_MOVE -> {
                    var newSelected: Point? = null
                    for (i in 0..boardSize - 1) {
                        for (j in 0..boardSize - 1) {
                            val point = locations[i][j]
                            if (Math.hypot((point.x - x).toDouble(), (point.y - y).toDouble()) < hexagonCRadius) {
                                newSelected = Point(i, j)
                            }
                        }
                    }
                    if (selectedHex == null || selectedHex != newSelected) {
                        selectedHex = newSelected
                        invalidate()
                    }
                }
                MotionEvent.ACTION_UP -> if (selectedHex != null) {
                    selectActionListener.onActionSelected(HexAction.valueOf(selectedHex!!.x, selectedHex!!.y))
                    selectedHex = null
                    invalidate()
                }
            }
            return true
        }
        return false
    }

    override fun onSizeChanged(width: Int, height: Int, oldWidth: Int, oldHeight: Int) {
        super.onSizeChanged(width, height, oldWidth, oldHeight)
        val paddingLeft = getPaddingLeft()
        val paddingTop = getPaddingTop()
        val paddingRight = getPaddingRight()
        val paddingBottom = getPaddingBottom()
        contentWidth = width - paddingLeft - paddingRight
        contentHeight = height - paddingTop - paddingBottom

        val maxHexWidth = contentWidth / (0.75f * boardSize + 0.25f)
        val maxHexHeight = ((2 * contentHeight) / (3 * boardSize - 1)).toFloat()
        hexagonRadius = Math.min(maxHexWidth, maxHexHeight / PathUtils.HEXAGON_SHORT_RADIUS) / 2
        hexagonCRadius = PathUtils.HEXAGON_SHORT_RADIUS * hexagonRadius
        val boardWidth = hexagonRadius * (1.5f * boardSize + 0.5f)
        val boardHeight = hexagonCRadius * (3 * boardSize - 1)
        val xPadding = (width.toFloat() - boardWidth) / 2
        val yPadding = (height.toFloat() - boardHeight) / 2
        val yAdjust = hexagonCRadius * (boardSize - 1)
        hexagon = PathUtils.getHexagon(hexagonRadius)
        lineWidth = hexagonRadius * LINE_WIDTH_RATIO

        val x0 = xPadding + hexagonRadius
        val y0 = yAdjust + yPadding + (PathUtils.HEXAGON_SHORT_RADIUS * hexagonRadius)
        for (i in 0..boardSize - 1) {
            for (j in 0..boardSize - 1) {
                locations[i][j] = PointF(x0 + i * hexagonRadius * 1.5f, y0 - i * hexagonCRadius + j * 2 * hexagonCRadius)
            }
        }
    }

    private fun getCorner(location: PointF, xChange: Int, yChange: Int): PointF {
        return PointF((location.x + xChange * PathUtils.HEXAGON_HALF_RADIUS * hexagonRadius), (location.y + yChange * PathUtils.HEXAGON_SHORT_RADIUS * hexagonRadius))
    }

    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)
        val matrix = Matrix()
        //draw board outer edges
        paint.setStyle(Paint.Style.FILL)
        paint.setStrokeWidth(lineWidth)
        paint.setColor(paletteColors.get(AGENT1_COLOR_VALUE))
        var p1 = getCorner(locations[0][0], -2, 0)
        var p2 = locations[0][boardSize - 1]
        canvas.drawRect(p1.x, p1.y, p2.x, p2.y, paint)
        p1 = getCorner(locations[boardSize - 1][0], 2, 0)
        p2 = locations[boardSize - 1][boardSize - 1]
        canvas.drawRect(p2.x, p1.y, p1.x, p2.y, paint)

        paint.setColor(paletteColors.get(AGENT2_COLOR_VALUE))
        p1 = getCorner(locations[0][0], -1, -1)
        p2 = getCorner(locations[boardSize - 1][0], -1, -1)
        var p3 = locations[boardSize - 1][0]
        var p4 = locations[0][0]
        path.reset()
        path.moveTo(p1.x, p1.y)
        path.lineTo(p2.x, p2.y)
        path.lineTo(p3.x, p3.y)
        path.lineTo(p4.x, p4.y)
        path.close()
        canvas.drawPath(path, paint)
        p1 = getCorner(locations[0][boardSize - 1], 1, 1)
        p2 = getCorner(locations[boardSize - 1][boardSize - 1], 1, 1)
        p3 = locations[boardSize - 1][boardSize - 1]
        p4 = locations[0][boardSize - 1]
        path.reset()
        path.moveTo(p1.x, p1.y)
        path.lineTo(p2.x, p2.y)
        path.lineTo(p3.x, p3.y)
        path.lineTo(p4.x, p4.y)
        path.close()
        canvas.drawPath(path, paint)

        // draw board background
        paint.setStyle(Paint.Style.FILL)
        for (i in 0..boardSize - 1) {
            for (j in 0..boardSize - 1) {
                paint.setColor(paletteColors.get(locationColors[i][j].toInt()))
                val temp = Path()
                val point = locations[i][j]
                matrix.reset()
                matrix.postTranslate(point.x - hexagonRadius, point.y - hexagonCRadius)
                hexagon!!.transform(matrix, temp)
                canvas.drawPath(temp, paint)
            }
        }
        // draw board inner edges
        paint.setStyle(Paint.Style.STROKE)
        paint.setColor(boardOutlineColor)
        //paint.setStrokeWidth(lineWidth);
        for (row in locations!!) {
            for (point in row) {
                val temp = Path()
                matrix.reset()
                matrix.postTranslate(point.x - hexagonRadius, point.y - hexagonCRadius)
                hexagon!!.transform(matrix, temp)
                canvas.drawPath(temp, paint)
            }
        }
        // draw selection lines
        if (selectedHex != null) {
            val y = selectedHex!!.y
            val x = selectedHex!!.x
            paint.setStyle(Paint.Style.STROKE)
            paint.setColor(lineColor)
            paint.setStrokeWidth(4.toFloat())
            val p1i = locations[x][0]
            val p1f = locations[x][locations.size() - 1]
            val p2i = locations[0][y]
            val p2f = locations[locations.size() - 1][y]
            canvas.drawLine(p1i.x, p1i.y, p1f.x, p1f.y, paint)
            canvas.drawLine(p2i.x, p2i.y, p2f.x, p2f.y, paint)
        }
    }
}